panic on consensus message decoding

Privnet setup:
```
INFO[1369] blockchain persist completed                  blockHeight=364506 headerHeight=594545 persistedBlocks=200 persistedKeys=601 took=1.297508ms
INFO[1370] blockchain persist completed                  blockHeight=364906 headerHeight=594545 persistedBlocks=400 persistedKeys=1201 took=2.520318ms
INFO[1371] blockchain persist completed                  blockHeight=365106 headerHeight=594545 persistedBlocks=200 persistedKeys=601 took=1.35826ms
INFO[1372] blockchain persist completed                  blockHeight=365506 headerHeight=594545 persistedBlocks=400 persistedKeys=1201 took=2.461972ms
panic: reflect.MakeSlice: negative len

goroutine 873 [running]:
reflect.MakeSlice(0xcfbaa0, 0xabfaa0, 0xa14db2aa8aa5a93a, 0xa14db2aa8aa5a93a, 0xabfaa0, 0x197, 0xc0000a0c48)
        /usr/lib64/go/1.12/src/reflect/value.go:2195 +0x1c4
github.com/CityOfZion/neo-go/pkg/io.(*BinReader).ReadArray(0xc001a15120, 0xaa8280, 0xc000144910)
        /home/rik/dev/neo-go/pkg/io/binaryReader.go:59 +0x1f9
github.com/CityOfZion/neo-go/pkg/consensus.(*prepareRequest).DecodeBinary(0xc000144900, 0xc001a15120)
        /home/rik/dev/neo-go/pkg/consensus/prepare_request.go:32 +0x9c
github.com/CityOfZion/neo-go/pkg/consensus.(*message).DecodeBinary(0xc0001acc00, 0xc001a15120)
        /home/rik/dev/neo-go/pkg/consensus/payload.go:134 +0x146
github.com/CityOfZion/neo-go/pkg/consensus.(*Payload).DecodeBinaryUnsigned(0xc0001acc00, 0xc001a150c0)
        /home/rik/dev/neo-go/pkg/consensus/payload.go:77 +0x151
github.com/CityOfZion/neo-go/pkg/consensus.(*Payload).DecodeBinary(0xc0001acc00, 0xc001a150c0)
        /home/rik/dev/neo-go/pkg/consensus/payload.go:90 +0x39
github.com/CityOfZion/neo-go/pkg/network.(*Message).decodePayload(0xc00349bf50, 0xc0000a0f78, 0xcefda0, 0x1255fc0)
        /home/rik/dev/neo-go/pkg/network/message.go:203 +0x224
github.com/CityOfZion/neo-go/pkg/network.(*Message).Decode(0xc00349bf50, 0xc0000a0f78, 0xc0034b0230, 0xc00349be60)
        /home/rik/dev/neo-go/pkg/network/message.go:163 +0x8c
github.com/CityOfZion/neo-go/pkg/network.(*TCPTransport).handleConn(0xc0001dad50, 0xcf13c0, 0xc0033be0c8)
        /home/rik/dev/neo-go/pkg/network/tcp_transport.go:84 +0x156
created by github.com/CityOfZion/neo-go/pkg/network.(*TCPTransport).Accept
        /home/rik/dev/neo-go/pkg/network/tcp_transport.go:59 +0x1a3
```

--- a/pkg/io/binaryReader.go
+++ b/pkg/io/binaryReader.go
@@ -3,9 +3,14 @@ package io
 import (
 	"bytes"
 	"encoding/binary"
+	"fmt"
 	"io"
 	"reflect"
 )
+
+// maxArraySize is a maximums size of an array which can be decoded.
+// It is taken from https://github.com/neo-project/neo/blob/master/neo/IO/Helper.cs#L130
+const maxArraySize = 0x1000000
 
 // BinReader is a convenient wrapper around a io.Reader and err object.
 // Used to simplify error handling when reading into a struct with many fields.
@@ -36,7 +41,7 @@ func (r *BinReader) ReadLE(v interface{}) {
 
 // ReadArray reads array into value which must be
 // a pointer to a slice.
-func (r *BinReader) ReadArray(t interface{}) {
+func (r *BinReader) ReadArray(t interface{}, maxSize ...int) {
 	value := reflect.ValueOf(t)
 	if value.Kind() != reflect.Ptr || value.Elem().Kind() != reflect.Slice {
 		panic(value.Type().String() + " is not a pointer to a slice")
@@ -55,7 +60,18 @@ func (r *BinReader) ReadArray(t interface{}) {
 		return
 	}
 
-	l := int(r.ReadVarUint())
+	ms := maxArraySize
+	if len(maxSize) != 0 {
+		ms = maxSize[0]
+	}
+
+	lu := r.ReadVarUint()
+	if lu > uint64(ms) {
+		r.Err = fmt.Errorf("array is too big (%d)", lu)
+		return
+	}
+
+	l := int(lu)
 	arr := reflect.MakeSlice(sliceType, l, l)
 
 	for i := 0; i < l; i++ {

--- a/pkg/io/binaryrw_test.go
+++ b/pkg/io/binaryrw_test.go
@@ -266,6 +266,17 @@ func TestBinReader_ReadArray(t *testing.T) {
 	require.NoError(t, r.Err)
 	require.Equal(t, elems, arrVal)
 
+	r = NewBinReaderFromBuf(data)
+	arrVal = []testSerializable{}
+	r.ReadArray(&arrVal, 3)
+	require.NoError(t, r.Err)
+	require.Equal(t, elems, arrVal)
+
+	r = NewBinReaderFromBuf(data)
+	arrVal = []testSerializable{}
+	r.ReadArray(&arrVal, 2)
+	require.Error(t, r.Err)
+
 	r = NewBinReaderFromBuf([]byte{0})
 	r.ReadArray(&arrVal)
 	require.NoError(t, r.Err)

